apis/v1beta1/validation/gateway.go

```diff
@@ -117,3 +119,41 @@ func ValidateTLSCertificateRefs(listeners []gatewayv1b1.Listener, path *field.Pa
 	}
 	return errs
 }
+
+
+// validateHostnameProtocolPort validates listener protocols or ports
+// must be different for the same hostname
+func ValidateHostnameProtocolPort(listeners []gatewayv1b1.Listener, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	hostnameProtocolMap := make(map[gatewayv1b1.Hostname][]gatewayv1b1.ProtocolType)
+	hostnamePortMap := make(map[gatewayv1b1.Hostname][]gatewayv1b1.PortNumber)
```

Comment:
Hey @gyohuangxin, I think there's a simpler way to do this. All that we really want to do is ensure that the combination of port, protocol, and name are unique. So for example, the following would be valid:

```yaml
- port: 80
  hostname: foo.com
  protocol: HTTP
- port: 443
  hostname: foo.com
  protocol: HTTP
```

But the following would not be valid:

```yaml
- port: 80
  hostname: foo.com
  protocol: HTTP
- port: 80
  hostname: foo.com
  protocol: HTTP
```

I think what you have is slightly different than that, requiring that for the same hostname (foo.com), both port _and_ protocol must be different for each listener. The spec as it's written today only requires that one of them is different.

A simple way to accomplish this would be to use something like `fmt.Sprintf("%s:%s:%d", protocol, hostname, port)` for each listener and then use k8s sets util to see if the value for that listener duplicated any listeners you'd already looked at within the same Gateway. Hopefully that makes sense, sorry for the delay on this!

Reply:
should this webhook attempt to reject one of these tuples (80, http, *.com) and (80, http, *.foo.com) :) ?

Reply:
@arkodg  I think it will very hard to accomplish. How would you decide if regexA (*.com in this case) is overlapping with regexB(*foo.com).

Reply:
yeah its hard, but I wanted to highlight the case.
Since the API calls out any regex match to be implementation specific, even adding language in the API for this might be tricky
cc @robscott